### DISCUSSION TITLE
How to Save React Flow Diagram as an Image?

### DISCUSSION THREAD

**Discussion by gpalrepo** [xyflow/xyflow#1061] [Q&A] [answered]
Hi, I want to save the react flow diagram as a png file. Wondering if anyone has suggestions? Thanks!

**Comment by Rudrava** [comment]
Hi @gpalrepo i was also trying to implement something like that can you please share any suggestion if you have made any progress in that !

**Reply by Matnard** [reply]
I did something similar recently using https://www.npmjs.com/package/dom-to-image

**Reply by Matnard** [reply]
https://codesandbox.io/s/spike-graph-export-bc9pu?file=/src/App.js
https://bc9pu.csb.app/

**Reply by Rudrava** [reply]
oh thank you so much this was a great help !!!

**Reply by moklick** [reply]
you could also try https://html2canvas.hertzen.com/ or for more advanced use cases render the flow on the server with puppeteer.

**Reply by prakash388** [reply]
Hi Team, I tried multiple ways but I can't able to. Please help me on this.

Note: My nodes contains images.  E.g Each node has 3 sections. First one Image section, second section is title of the node and third is also Image section. It should capture the images also

**Reply by Rudrava** [reply]
is it that only your images are not showing up on the image ! ?

**Reply by amd-99** [reply]
### Solution for Export react-flow-renderer flow as PDF
Here, I found solution for Export react flow as pdf  or SVG with out quality compromise ,I mentioned all steps and attached example

step 1. Need to convert  flow into SVG
step 2. After that render SVG into new tab or new window
step 3. And print simply with window.print() and save as pdf.


Example for ReactJS :-
            `      

              async function exportFlow (){
              
             //for get whole chart print need to fit-view

             reactFlowInstance.fitView( );

             function filter (node: { tagName: string; }) {
             return (node.tagName !== 'i');
              }

         //get only printable elements by className
          let elements = document.getElementsByClassName('react-flow__renderer')[0] as any;


      //converts html to svg (here I used htmlToImage library and its converts with foreign object of svg  , so 
       need to open in new window to render)
         htmlToImage.toSvg(elements, { filter: filter }).then(async(svgContent: any)=>{
        
           const svgElement =  await decodeURIComponent(svgContent.replace("data:image/svg+xml;charset=utf-8,",''").trim();

       const newWindow:any = open();
        newWindow.document.write(
           `<html>
            <head>
            <title>Flow.pdf</title>
            <style>
            @page {
              size: A4 landscape !important;
              margin:0 !important;
            }
            @media print {
            * {
            -webkit-print-color-adjust: exact !important;   /* Chrome, Safari */
            color-adjust: exact !important;                 /*Firefox*/
             }
            }
            </style>
            </head>
            <body style="margin:60px 32px 32px 32px ">
                    ${svgElement}
            <script>
            window.print();
            window.close();
            </script>
            </body>
            </html>`)
      })
    }`

**Reply by zayonoah** [reply]
@Amd-99 Thank you for posting this. I was finally able to get a solid vectorized pdf working. 

For reference, this is the code I ended up with. Not shown: I also had to set the pdf height and width to the react flow parent container to get it the same size needed for the pdf. After exporting is complete, I change the parent height and width back


````
async function exportDiagram() {
        try {
            // Needs these sleep to make sure flow is completely rendered
            await sleep(1000);
            // Remove animation from edges since animation wont be shown in pdf
            if (reactFlowInstance) {
                reactFlowInstance.setEdges((edges) =>
                    edges.map((edge) => {
                        if (edge) {
                            edge.animated = false;
                            edge.markerEnd = {
                                type: 'arrow',
                                width: 15,
                                height: 15
                            };
                        }

                        return edge;
                    })
                );
                reactFlowInstance.fitView();
            }
            // Again having issues without small sleep. Should be removable though.
            await sleep(1000);
            // Get the DOM element
            let elements = document.getElementsByClassName('react-flow-exporting')[0];
            // Convert to SVG
            const svgContent = await htmlToImage.toSvg(elements, { filter: filter });
            const svgElement =  await decodeURIComponent(svgContent.replace("data:image/svg+xml;charset=utf-8,", "").trim());
            // Open new window
            const newWindow = open();
            // Safer version of document.write
            document.write=function(s){
                var scripts = document.getElementsByTagName('script');
                var lastScript = scripts[scripts.length-1];
                lastScript.insertAdjacentHTML("beforebegin", s);
            }
            // Write our page content to the newly opened page
            newWindow.document.write(
                `<html>
                    <head>
                        <title>React Flow PDF</title>
                        <style>
                            body {
                                width: ${width};
                                height: ${height};
                                margin: auto
                            }
                            .container {
                                background: #393D43;
                                text-align: center;
                                height: 100%;
                                width: 100%;
                            }
                            
                            @page {
                                size: 29.7cm, 21cm
                                margin:0 !important;
                            }
                            @media print {
                                * {
                                    -webkit-print-color-adjust: exact !important;
                                    color-adjust: exact !important;
                                }
                                .container {
                                    background: none;
                                }
                            }
                        </style>
                    </head>
                    <body>
                        <div class='container'>
                            <div class='svg-container'>
                                ${svgElement}
                            </div>
                            
                            <script>
                                document.close();
                                window.print();
                            </script>
                        </div>
                    </body>
                </html>`
            )
        } catch (error) {
            console.log(error);
        }
    }````

**Reply by nadiTime** [reply]
link to codesandbox? 🙏

**Comment by moklick** [comment, marked as answer]
I've created a ["Download Image" example](https://reactflow.dev/docs/examples/misc/download-image/) that uses [html-to-image](https://github.com/bubkoo/html-to-image).

**Reply by rahatzamancse** [reply]
Nice work! I actually prefer to have the download button inside the `<Control>` tag instead. Here is your refactored working code for this:

```tsx
<ReactFlow
    nodes={nodes}
    edges={edges}
    ref={flowRef}
>
    <Controls >
        <ControlButton onClick={() => {
            if(flowRef.current === null) return
            toPng(flowRef.current, {
                filter: node => !(
                    node?.classList?.contains('react-flow__minimap') ||
                    node?.classList?.contains('react-flow__controls')
                ),
            }).then(dataUrl => {
                const a = document.createElement('a');
                a.setAttribute('download', 'reactflow.png');
                a.setAttribute('href', dataUrl);
                a.click();
            });
        }}>
            <img src="assets/export.png" alt="Export" width="16px" height="16px" />
        </ControlButton>
    </Controls>
    <Background />
</ReactFlow>
```

**Comment by yavor-grancharov** [comment]
@moklick Hi, I am currently trying to download whole flow as png or pdf but wasnt able to make it work in a way that everything is inside even nodes outside of viewport. I am using html-to-image to convert reactflow ref to canvas. My lastest try contains code below:
```
const downloadCurrentWorkflow = React.useCallback(() => {
    if (reactflowRef.current === null) return;
    reactFlowInstance?.fitView();
    toCanvas(reactflowRef.current, {
      filter: (node) =>
        !(
          node?.classList?.contains('react-flow__minimap') ||
          node?.classList?.contains('react-flow__controls')
        ),
    })
      .then((canvas) => {
        const imgData = canvas.toDataURL('image/png');
        const imgWidth = 210;
        const pageHeight = 295;
        const imgHeight = (canvas.height * imgWidth) / canvas.width;
        let heightLeft = imgHeight;
        let pos = 0;
        const pdf = new jsPDF('p', 'mm');
        pdf.addImage(imgData, 'PNG', 0, pos, imgWidth, imgHeight);
        heightLeft -= pageHeight;
        while (heightLeft >= 0) {
          pos = heightLeft - imgHeight;
          pdf.addPage();
          pdf.addImage(imgData, 'PNG', 0, pos, imgWidth, imgHeight);
          heightLeft -= pageHeight;
        }
        pdf.save('reactflow.pdf');
      })
      .catch((err) => {
        console.warn('Downloading error: ', err);
      });
  }, [reactflowRef]);

  return (
    <ReactFlow
      nodes={nodes}
      edges={edges}
      nodeTypes={nodeTypes}
      onNodesChange={onNodesChange}
      onEdgesChange={onEdgesChange}
      onEdgeUpdate={onEdgeUpdate}
      onEdgeUpdateStart={onEdgeUpdateStart}
      onEdgeUpdateEnd={onEdgeUpdateEnd}
      onConnect={onConnect}
      onConnectStart={onConnectStart}
      onNodeDoubleClick={(_, node) => onClick(node, 'NODE')}
      onEdgeDoubleClick={(_, edge) => onClick(edge, 'EDGE')}
      onInit={(rfIs) => setReactFlowInstance(rfIs)}
      onDrop={onDrop}
      onDragOver={onDragOver}
      onNodesDelete={onNodesDeleteHandler}
      ref={reactflowRef}
    >
      <Background />
      <Controls>
        <ControlButton onClick={downloadCurrentWorkflow}>
          <Icon icon="download" />
        </ControlButton>
      </Controls>
    </ReactFlow>
  );
```
Is it possible to make this work? Thanks!

**Reply by moklick** [reply]
Hey @yavor-grancharov 

you can overwrite the current transform to reset it. Here is an example https://codesandbox.io/s/image-fullscreen-936g03?file=/DownloadButton.js:356-425 . You can calculate the transform for the nodes with getRectOfNodes -> getTransformForBounds https://reactflow.dev/docs/api/graph-util-functions/#getrectofnodes

**Comment by msholty-fd** [comment]
I didn't see anyone come out and fully lay out an example that works that also captures the entire viewport in case you have a big reactflow instance. Here's what worked for me:

```tsx
import { useState, useEffect } from "react";
import { getRectOfNodes, getTransformForBounds, useReactFlow } from "reactflow";
import { toPng } from "html-to-image";

function DownloadFlow() {
  const [previewImage, setPreviewImage] = useState<string | null>(null);
  const { getNodes } = useReactFlow();

  useEffect(() => {
      const nodesBounds = getRectOfNodes(getNodes());
      const imageWidth = nodesBounds.width + 200;
      const imageHeight = nodesBounds.height + 200;
      const transform = getTransformForBounds(
        nodesBounds,
        imageWidth,
        imageHeight,
        0.5,
        2,
      );

      toPng(document.querySelector(".react-flow__viewport") as HTMLElement, {
        backgroundColor: Colors.neutral200,
        width: imageWidth,
        height: imageHeight,
        style: {
          width: String(imageWidth),
          height: String(imageHeight),
          transform: `translate(${transform[0]}px, ${transform[1]}px) scale(${transform[2]})`,
        },
      }).then((dataUrl) => {
        setPreviewImage(dataUrl);
      });
  }, [getNodes]);

  function handleActualDownload() {
    if (previewImage) {
      downloadImage(previewImage, name);
      emitEventOnDownload();
      onClosed();
    }
  }

  return <button onClick={handleActualDownload}>Download</button>;
}

function downloadImage(dataUrl: string, name?: string) {
  const now = new Date();
  const dateStr = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(
    2,
    "0",
  )}-${String(now.getDate()).padStart(2, "0")}`;

  const timeStr = `${String(now.getHours()).padStart(2, "0")}${String(
    now.getMinutes(),
  ).padStart(2, "0")}${String(now.getSeconds()).padStart(2, "0")}`;

  const fileNamePrefix = name ? `${name}_` : "reactflow_";
  const fileName = `${fileNamePrefix}${dateStr}_${timeStr}.png`;

  const a = document.createElement("a");
  a.setAttribute("download", fileName);
  a.setAttribute("href", dataUrl);
  a.click();
}

**Reply by useyourillusions** [reply]
That's was incredibly helpful, thank you!

**Comment by cini9** [comment]
Hey guys, I came across this discussion when trying to debug my download functionality as well. As a quick heads up, it appears that using the nodeOrigin prop on the ReactFlow component causes the downloaded image to be offset and for a portion of the left side and the top of the image to be cropped. 

My ReactFlow component :

```typescript
<ReactFlow
  nodes={nodes}
  edges={edges}
  ...
  nodeOrigin={[0.5, 0.5]}
  ...
  >
  <Background />
  
  <Controls />
</ReactFlow>
``` 

And my download functionality :

```typescript
const screenshot = () => {
  const nodesBounds = getNodesBounds(getNodes())
  const imageWidth = 1024
  const imageHeight = 768
  
  const transform = getViewportForBounds(
    nodesBounds,
    imageWidth,
    imageHeight,
    0,
    2
  )
  
  toPng(document.querySelector('.react-flow__viewport') as HTMLElement, {
    filter: node =>
      !(
        node?.classList?.contains('react-flow__minimap') ||
        node?.classList?.contains('react-flow__controls')
      ),
    backgroundColor: '#fff',
    width: imageWidth,
    height: imageHeight,
    style: {
      transform: `translate(${transform.x}px, ${transform.y}px) scale(${transform.zoom})`,
    },
  }).then(dataUrl => {
    const a = document.createElement('a')
    
    a.setAttribute('download', 'reactflow.png')
    a.setAttribute('href', dataUrl)
    a.click()
  })
}
```

Gives this cropped image :
![reactflow](https://github.com/xyflow/xyflow/assets/56114997/39aa8d05-e68c-4c36-aaee-8910c3a21a2a)

If i comment out nodeOrigin :
![reactflow (1)](https://github.com/xyflow/xyflow/assets/56114997/2296cfbe-1e1c-42f6-bedc-88076619fa70)

It feels like getNodesBounds or getViewportForBounds (or both?) doesn't take the nodeOrigin into account. Can anybody confirm this ? Should this require a fix ? Or does anybody have an idea for a workaround ?

Much appreciated !

**Comment by nsarwar123** [comment]
**How can we take image of a group Node only instead of full flow?**

As in [download-image](https://reactflow.dev/examples/misc/download-image) example its using **.react-flow__viewport** class to take an image of all nodes in the flow.
In may case I have multiple group nodes _i.e pages(A4)_ and in every node, there are some child nodes. How can I make an image of any single group node including its child nodes (like, getting an image of any page from multiple pages).

![image](https://github.com/user-attachments/assets/ac4ee1e7-fb99-43a4-8fb9-19123a5b56db)

**Comment by knackstedt** [comment]
I solved this using [dom-to-svg](https://www.npmjs.com/package/dom-to-svg) instead of html-to-image because I was having a lot of issues with html-to-image not rendering edges and the output SVG had a ton of unnecessary attributes that made the output file ~20MB where the PNG variant was ~400K.

Sample:
https://github.com/knackstedt/surrealist/blob/0a9bc5589353d3172e497f729a9bda87ac3eb573/src/screens/surrealist/views/designer/TableGraphPane/helpers.tsx#L361

**Comment by jmhache** [comment]
You can also try [SnapDOM](https://github.com/zumerlab/snapdom) for this. It captures any DOM element to SVG, PNG, JPEG or Blob, handles complex CSS well, and is a lot faster than html2canvas or html-to-image. Works nicely with dynamic structures like flow diagrams.
